### PR TITLE
Refactor Segment Granularity

### PR DESCRIPTION
**Pull request by praveev**
The goal of this PR is to remove the `Granularity` enum which was used earlier and replace it with something more sophisticated that takes timezones into account.  Hence the new `SegmentGranularity`. The new class is modeled after `QueryGranularity`.

The earlier timezone support I created had some caveats https://github.com/druid-io/druid/pull/3528 Timezone wasn't respected for real time and also for batch ingestion without intervals. With this change, timezone can now be enforced with segment granularity for both batch ingestion and real time ingestion. Only batch ingestion has been tested.

`Granularity` had plenty of unused methods. Those have been thrown away and all its tests have been ported for `SegmentGranularity`. 

Core files that were added/modified:
1. SegmentGranularity
2. PeriodSegmentGranularity
3. UniformGranularitySpec
4. PeriodGranularity
5. PeriodQueryGranularity

All the other files are mostly side effects of these changes.

### CODE REVIEW

**Comment by himanshug** [comment]
i reviewed it internally, :+1:

File: java-util/src/main/java/io/druid/java/util/common/granularity/PeriodGranularity.java [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+
+public class PeriodGranularity {
+    private final Period period;
+    private final Chronology chronology;
+    private final long origin;
+    private final boolean hasOrigin;
+
+    private final boolean isCompound;
+
+    public PeriodGranularity(Period period, DateTime origin, DateTimeZone tz)
```
Comment (fjy) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodGranularity.java]: is this class only for segments or queries?

If so, can we call this SegmentGranularity or QueryGranularity?

I prefer for there to be SegmentGranularity and QueryGranularity, and for us to get rid of the general class Granularity

**Comment by fjy** [comment]
I'm confused by this PR. Our goal has been to move towards a single concept of Granularity and now we have 3 granularities. Can this decision plz be justified?

**Comment by praveev** [comment]
@fjy I wasn't aware that there was an explicit set goal that we wanted to only have a single concept of Granularity. 

I had earlier sent a PR for https://github.com/druid-io/druid/pull/3528. For that, I had to work around the fact that Granularity enum inherently didn't support timezone. Early on, @cheddar suggested to solve this by taking note of how QueryGranularity supports `period`. But we didn't have the time to do this before the first PR. Now, I ended up refactoring by creating a `PeriodSegmentGranularity` and `PeriodQueryGranularity`. Both of them have some shared logic about truncation, increments and decrements for a `period`. That shared logic was then encapsulated in as `PeriodGranularity`.

**Comment by fjy** [comment]
@praveev I think for future developers and to describe Druid to folks, it'd be good to create a unified Granularity object that understands timezones and can be used for queries and ingestion. Can we do it in this PR?

**Comment by praveev** [comment]
Thanks @leventov. I've fixed the formatting. 

@himanshug @leventov I've also fixed an issue in the code that will cause `PeriodSegmentGranularity` to throw an exception in the constructor for custom periods. Please take a look  	088c990

@fjy thanks. I've begun working on consolidating segment and query granularity stuff on top of this PR. But it will take sometime for me to complete and test the changes thoroughly.

**Comment by leventov** [comment]
Please don't merge, I'm going to review this before the end of the week

**Comment by praveev** [comment]
Thanks @leventov. I've fixed the formatting.
@himanshug @leventov I found two issues since and have fixed them in 088c990:

1. The code for mapping a period to granularityType only worked if the period was in one of the pre-defined granularities (the ones from GranularityType). As a result, if we specified a custom period like say PT2S or P2M etc, the constructor would throw an exception b/c it wouldn’t be able to find an appropriate granularityType. The reason the mapping exists in the first place is so that we can get the appropriate formatters.
2. The custom serialization didn’t take into account custom periods. The logic around that serialization to retain old behavior was wrong.

@fjy thanks. I've begun working on consolidating segment and query granularity stuff on top of this PR. But it will take sometime for me to complete and test the changes thoroughly.

File: server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java [hunk 2]
```diff
@@ -46,34 +45,25 @@
...
   private final QueryGranularity queryGranularity;
   private final Boolean rollup;
-  private final String timezone;
 
   @JsonCreator
   public ArbitraryGranularitySpec(
       @JsonProperty("queryGranularity") QueryGranularity queryGranularity,
       @JsonProperty("rollup") Boolean rollup,
-      @JsonProperty("intervals") List<Interval> inputIntervals,
-      @JsonProperty("timezone") String timezone
```
Comment (leventov) [inline server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java]: Isn't this change backwards-incompatible?

File: server/src/main/java/io/druid/segment/indexing/granularity/UniformGranularitySpec.java [hunk 3]
```diff
@@ -28,67 +28,57 @@
...
-  private final String timezone;
 
   @JsonCreator
   public UniformGranularitySpec(
-      @JsonProperty("segmentGranularity") Granularity segmentGranularity,
+      @JsonProperty("segmentGranularity") SegmentGranularity segmentGranularity,
       @JsonProperty("queryGranularity") QueryGranularity queryGranularity,
       @JsonProperty("rollup") Boolean rollup,
-      @JsonProperty("intervals") List<Interval> inputIntervals,
-      @JsonProperty("timezone") String timezone
```
Comment (leventov) [inline server/src/main/java/io/druid/segment/indexing/granularity/UniformGranularitySpec.java]: Isn't this change backwards-incompatible?

File: java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+  @JsonCreator
+  public PeriodSegmentGranularity(
+      @JsonProperty("period") Period period,
+      @JsonProperty("origin") DateTime origin,
+      @JsonProperty("timeZone") DateTimeZone tz
+  )
+  {
+    this.periodGranularity = new PeriodGranularity(period, origin, tz);
+    this.granularityType = GranularityType.estimatedGranularityType(period);
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java]: Currently, segmentGranularity has to be one of the predefined values as per the Granularity enum i.e SECOND, MINUTE, FIVE_MINUTE, TEN_MINUTE etc 

With this PR, segmentGranularity can be represented with periods. But with one restriction, the periods can't be compound meaning 1hour5minutes etc. This is because compound periods makes it harder to match a "hive", "default", "lowerDefault" format representation. We don't have this format issue with custom periods that are not compound. For example: PT2S has the same format as PT1S. PT13M has the same format as PT1M.

So, that's why `GranularityType.estimatedGranularityType(period);` will throw `IAE` for compound periods. 

I have two tests for this: `testCustomPeriodToDate` and `testCustomNestedPeriodFail`.

File: server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java (hunk 2 above)
Comment (praveev) [inline server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java]: yes. this change is only in master. has not been released yet.

File: server/src/main/java/io/druid/segment/indexing/granularity/UniformGranularitySpec.java (hunk 3 above)
Comment (praveev) [inline server/src/main/java/io/druid/segment/indexing/granularity/UniformGranularitySpec.java]: yes. but this change is only in master. has not been released yet.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+        }
+      }
+
+      if (bs.cardinality() == 0 || bs.cardinality() > 1) {
+        throw new IAE("Granularity is not supported. [%s]", period);
+      }
+      else {
+        final int index = bs.nextSetBit(0);
+
+        if (index == 0) {
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: yes. its don't intentionally b/c granualrityType is mainly used to get the corresponding format. And format for 5, 10, 15min is not different than 1min.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+    }
+
+    // Note: This is only an estimate based on the values in period.
+    // This will not work for complicated periods that represent say 1 year 1 day
+    static GranularityType estimatedGranularityType(Period period) {
+      int[] vals = period.getValues();
+      BitSet bs = new BitSet();
+      for (int i = 0; i < vals.length; i++) {
+        if (vals[i] != 0) {
+          bs.set(i);
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: The BitSet exists to throw an `IAE` for compound periods. B/c compound periods don't have a matching "hive", "default" or "lowerDefault" format.

File: java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+      JsonGenerator jsonGenerator,
+      SerializerProvider serializerProvider,
+      TypeSerializer typeSerializer
+  ) throws IOException, JsonProcessingException
+  {
+    serialize(jsonGenerator, serializerProvider);
+  }
+
+  private boolean equalsPredefinedGranularities() {
+    final GranularityType[] values = GranularityType.values();
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java]: I thought about that. But keeping a list separately means that the list needs to be updated anytime new predefined granularities are added. This code can go away if we don't support rolling upgrade in 0.10.0.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java [hunk 8]
```diff
@@ -0,0 +1,450 @@
...
+      Pattern.compile("^.*dt=(\\d{4})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})(?:-(\\d{2})?)?)?)?)?)?/.*$");
+
+  public enum Formatter
+  {
+    DEFAULT,
+    HIVE,
+    LOWER_DEFAULT
+  }
+
+  public static final SegmentGranularity SECOND = SegmentGranularity.fromString("SECOND");
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: I'm guessing you're suggesting to directly call `PeriodSegmentGranualrity` here. 
This is done so for uniformity and maintainability. Jackson will use `fromString`. So if we change internals of this method in the future, then we'd also have to change it here.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java [hunk 9]
```diff
@@ -0,0 +1,450 @@
...
+   */
+  public final Interval bucket(DateTime t)
+  {
+    DateTime start = truncate(t);
+    return new Interval(start, increment(start));
+  }
+
+  // Only to create a mapping of the granularity and all the supported file patterns
+  // namely: default, lowerDefault and hive.
+  protected enum GranularityType
```
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: `granularitiesFinerThan` static method uses GranularityType. So 5, 10, 15min can't be removed easily from this enum.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java (hunk 8 above)
Comment (leventov) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: I mean that we are in SegmentGranularity.java and calling `SegmentGranularity.fromString()`. Couldn't it be just `fromString()`?

File: java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java (hunk 7 above)
Comment (leventov) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java]: With the currently suggested code, if a new predefined granularity added, the switch statement below should also be updated.

> This code can go away if we don't support rolling upgrade in 0.10.0.

I don't understand, could you please elaborate?

File: server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java (hunk 2 above)
Comment (leventov) [inline server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java]: Do you mean that another PR should be done before 0.10.0 to restore backwards compatibility?

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java (hunk 6 above)
Comment (leventov) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: If you mean `IAE` on line 218, the current logic is preserved by making a variable like `granularityType = null` outside of this loop, assinging `granularityType = GranularityType.YEAR` etc. in the "switch" below, then throwing `IAE` if `vals[i] != null` and `granularityType != null` (eq. to your `bs.cardinality() > 1`) and also throw `IAE` if `granularityType == null` after the loop, before return (eq. to your `bs.cardinality() == 0`).

File: java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    public String toString() {
+        return "{type=period, " +
+               "period=" + getPeriod() +
+               ", timeZone=" + getTimeZone() +
+               ", origin=" + getOrigin() +
+               '}';
+    }
+
+    @Override
+    public void serialize(JsonGenerator jsonGenerator, SerializerProvider serializerProvider) throws IOException, JsonProcessingException {
```
Comment (cheddar) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java]: "Not supporting rolling upgrade for 0.10.0" is not optional.  Supporting rolling updates is always required, even if we make backwards incompatible code changes.

Comment (cheddar) [inline java-util/src/main/java/io/druid/java/util/common/granularity/PeriodSegmentGranularity.java]: Btw, this method can also be a lot simpler.  You can just do something like

```
@JsonValue
public Object getJson() {
  if (getTimeZone() == DateTimeZone.UTC && getOrigin() == null) {
         return granularityType.toString();
  } else {
    Map<String, Object> retVal = Maps.newLinkedHashMap();
    retVal.put("type", "period");
    retVal.put("period", getPeriod());
    ...
    return retVal;
  }
}
```

File: server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java (hunk 2 above)
Comment (cheddar) [inline server/src/main/java/io/druid/segment/indexing/granularity/ArbitraryGranularitySpec.java]: No, his response is saying that it's backwards incompatible, but it's backwards incompatible with something that only exists in master, not in any released code.  So, it won't actually be backwards incompatible as far as the release is concerned.

**Comment by praveev** [comment]
@fjy I've created a PR to unify granularities #3850. Please take a look.

File: java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java (hunk 5 above)
Comment (praveev) [inline java-util/src/main/java/io/druid/java/util/common/granularity/SegmentGranularity.java]: I updated to return the correct GranularityType for 5, 10, 15mins in #3850

**Comment by cheddar** [comment]
Note, we are running this in production currently.
